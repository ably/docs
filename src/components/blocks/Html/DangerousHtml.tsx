import React from 'react';
import DOMPurify from 'dompurify';

export const DangerousHtml = ({ data }: { data: string }) => (
  <div
    dangerouslySetInnerHTML={{
      __html: DOMPurify.sanitize(data, {
        // The SVG and Math tags have been used in the past as attack vectors for mXSS,
        // but if we really need them should be safe enough to enable.
        // This is probably too cautious but we have no need for them at time of writing, so forbidding them is free.
        FORBID_TAGS: ['svg', 'math'],
      }),
    }}
  ></div>
);
