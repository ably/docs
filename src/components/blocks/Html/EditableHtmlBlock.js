import React from 'react';
import PropTypes from 'prop-types';

const EditableHtmlBlock = (Type) => {
  const InnerBlock = ({ data, attribs }) => <Type defaultValue={data} {...attribs} />;
  InnerBlock.propTypes = {
    data: PropTypes.string,
    attribs: PropTypes.object,
  };
  return InnerBlock;
};

export default EditableHtmlBlock;
