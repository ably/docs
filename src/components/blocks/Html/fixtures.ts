export const dtFixture = [
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: [
      {
        data: 'name',
        type: 'text',
        name: 'text',
      },
    ],
    type: 'tag',
    name: 'dt',
    attribs: {},
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: [
      {
        data: 'event name for the published message',
        type: 'text',
        name: 'text',
      },
      {
        data: [],
        type: 'tag',
        name: 'br',
        attribs: {},
      },
      {
        data: [
          {
            data: [
              {
                data: 'Type: ',
                type: 'text',
                name: 'text',
              },
              {
                data: [
                  {
                    data: 'String',
                    type: 'text',
                    name: 'text',
                  },
                ],
                type: 'tag',
                name: 'code',
                attribs: {},
              },
            ],
            type: 'tag',
            name: 'em',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'span',
        attribs: {
          lang: 'default',
        },
      },
      {
        data: [
          {
            data: [
              {
                data: 'Type: ',
                type: 'text',
                name: 'text',
              },
              {
                data: [
                  {
                    data: 'Unicode',
                    type: 'text',
                    name: 'text',
                  },
                ],
                type: 'tag',
                name: 'code',
                attribs: {},
              },
              {
                data: ' for Python 2, ',
                type: 'text',
                name: 'text',
              },
              {
                data: [
                  {
                    data: 'String',
                    type: 'text',
                    name: 'text',
                  },
                ],
                type: 'tag',
                name: 'code',
                attribs: {},
              },
              {
                data: ' for Python 3',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'em',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'span',
        attribs: {
          lang: 'python',
        },
      },
    ],
    type: 'tag',
    name: 'dd',
    attribs: {},
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: [
      {
        data: [
          {
            data: [
              {
                data: 'data',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'div',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'dt',
        attribs: {},
      },
      {
        data: [
          {
            data: 'data payload for the message. The supported payload types are unicode Strings, Dict, or List objects that can be serialized to ',
            type: 'text',
            name: 'text',
          },
          {
            data: [
              {
                data: 'JSON',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'span',
            attribs: {
              class: 'caps',
            },
          },
          {
            data: ' using ',
            type: 'text',
            name: 'text',
          },
          {
            data: [
              {
                data: 'json.dumps',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'code',
            attribs: {},
          },
          {
            data: ', binary data as ',
            type: 'text',
            name: 'text',
          },
          {
            data: [
              {
                data: 'bytearray',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'code',
            attribs: {},
          },
          {
            data: ' (in Python 3, ',
            type: 'text',
            name: 'text',
          },
          {
            data: [
              {
                data: 'bytes',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'code',
            attribs: {},
          },
          {
            data: ' also works), and None.',
            type: 'text',
            name: 'text',
          },
          {
            data: [],
            type: 'tag',
            name: 'br',
            attribs: {},
          },
          {
            data: [
              {
                data: 'Type: ',
                type: 'text',
                name: 'text',
              },
              {
                data: [
                  {
                    data: 'Object',
                    type: 'text',
                    name: 'text',
                  },
                ],
                type: 'tag',
                name: 'code',
                attribs: {},
              },
            ],
            type: 'tag',
            name: 'em',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'dd',
        attribs: {},
      },
    ],
    type: 'tag',
    name: 'div',
    attribs: {
      lang: 'python',
    },
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: [
      {
        data: [
          {
            data: [
              {
                data: 'data',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'div',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'dt',
        attribs: {},
      },
      {
        data: [
          {
            data: 'data payload for the message. The supported payload types are String, Map, List, and null.',
            type: 'text',
            name: 'text',
          },
          {
            data: [],
            type: 'tag',
            name: 'br',
            attribs: {},
          },
          {
            data: [
              {
                data: 'Type: ',
                type: 'text',
                name: 'text',
              },
              {
                data: [
                  {
                    data: 'Object',
                    type: 'text',
                    name: 'text',
                  },
                ],
                type: 'tag',
                name: 'code',
                attribs: {},
              },
            ],
            type: 'tag',
            name: 'em',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'dd',
        attribs: {},
      },
    ],
    type: 'tag',
    name: 'div',
    attribs: {
      lang: 'flutter',
    },
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: [
      {
        data: [
          {
            data: [
              {
                data: 'message',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'div',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'dt',
        attribs: {},
      },
      {
        data: [
          {
            data: 'payload for the message as a ',
            type: 'text',
            name: 'text',
          },
          {
            data: [
              {
                data: 'Message',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'code',
            attribs: {},
          },
          {
            data: ':#message object.',
            type: 'text',
            name: 'text',
          },
          {
            data: [],
            type: 'tag',
            name: 'br',
            attribs: {},
          },
          {
            data: [
              {
                data: 'Type: ',
                type: 'text',
                name: 'text',
              },
              {
                data: [
                  {
                    data: 'Message',
                    type: 'text',
                    name: 'text',
                  },
                ],
                type: 'tag',
                name: 'code',
                attribs: {},
              },
              {
                data: ':#message',
                type: 'text',
                name: 'text',
              },
            ],
            type: 'tag',
            name: 'em',
            attribs: {},
          },
        ],
        type: 'tag',
        name: 'dd',
        attribs: {},
      },
    ],
    type: 'tag',
    name: 'div',
    attribs: {
      lang: 'flutter',
    },
  },
  {
    data: '\n\t',
    type: 'text',
    name: 'text',
  },
  {
    data: '\n',
    type: 'text',
    name: 'text',
  },
];
