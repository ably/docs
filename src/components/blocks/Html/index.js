import React from 'react';
import PropTypes from 'prop-types';
import DOMPurify from 'dompurify';
import blocksFromData from '../blocks-from-data';
import { isArray } from 'lodash';
import ConditionalChildrenLanguageDisplay from '../wrappers/ConditionalChildrenLanguageDisplay';

// This is not to be used in production.
const DangerousHtml = ({ data }) => (
  <div
    dangerouslySetInnerHTML={{
      __html: DOMPurify.sanitize(data, {
        // The SVG and Math tags have been used in the past as attack vectors for mXSS,
        // but if we really need them should be safe enough to enable.
        // This is probably too cautious but we have no need for them at time of writing, so forbidding them is free.
        FORBID_TAGS: ['svg', 'math'],
      }),
    }}
  ></div>
);

DangerousHtml.propTypes = {
  data: PropTypes.string,
};

const Html = ({ data }) =>
  isArray(data) ? (
    <ConditionalChildrenLanguageDisplay>{blocksFromData(data)}</ConditionalChildrenLanguageDisplay>
  ) : (
    <>{data}</>
  );

Html.propTypes = {
  data: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
};

export { DangerousHtml };
export default Html;
