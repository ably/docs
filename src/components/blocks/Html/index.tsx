import React, { FunctionComponent } from 'react';
import { isArray } from 'lodash';
import { HtmlComponentProps, ValidReactElement } from 'src/components/html-component-props';
import { ArticleTypeContext } from '../../../contexts/article-type-context';
import { ARTICLE_TYPES } from '../../../../data/transform/constants';
import { defaultBlocksFromData } from '../create-blocks/default-blocks-from-data';
import { apiReferenceBlocksFromData } from '../create-blocks/api-reference-blocks-from-data';
import ConditionalChildrenLanguageDisplay from '../wrappers/ConditionalChildrenLanguageDisplay';

const Html = ({
  data,
  BlockWrapper = ConditionalChildrenLanguageDisplay as FunctionComponent,
}: {
  data: HtmlComponentProps<ValidReactElement>[] | string | null;
  BlockWrapper?: React.FunctionComponent;
}) => {
  if (isArray(data)) {
    return (
      <ArticleTypeContext.Consumer>
        {(articleType) => {
          const blockCreator =
            articleType === ARTICLE_TYPES.apiReference ? apiReferenceBlocksFromData : defaultBlocksFromData;
          return <BlockWrapper>{blockCreator(data)}</BlockWrapper>;
        }}
      </ArticleTypeContext.Consumer>
    );
  }

  return <>{data}</>;
};

export default Html;
