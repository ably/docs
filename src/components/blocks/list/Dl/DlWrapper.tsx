import React, { Children, FunctionComponent, useContext } from 'react';
import PageLanguageContext from '../../../../contexts/page-language-context';
import { DEFAULT_LANGUAGE, DEFAULT_PREFERRED_LANGUAGE } from '../../../../../data/createPages/constants';

export const DlWrapper: FunctionComponent<any> = ({ children }) => {
  const language = useContext(PageLanguageContext);

  return Children.map(children, (child) => {
    const { attribs } = child?.props;

    if (attribs?.lang) {
      if (
        (language === DEFAULT_LANGUAGE && attribs?.lang === DEFAULT_PREFERRED_LANGUAGE) ||
        attribs?.lang === language
      ) {
        return child;
      }
      return null;
    }
    return child;
  });
};
