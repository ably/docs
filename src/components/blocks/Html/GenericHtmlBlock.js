import React from 'react';
import Html from '.';

const GenericHtmlBlock = (Type) => {
  const InnerBlock = ({ data, attribs = {} }) => (
    <Type {...attribs}>
      <Html data={data} />
    </Type>
  );
  return InnerBlock;
};

export default GenericHtmlBlock;
