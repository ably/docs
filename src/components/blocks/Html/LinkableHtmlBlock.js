import React from 'react';
import PropTypes from 'prop-types';
import Html from '.';
import CopyLink from '../wrappers/CopyLink';

const LinkableHtmlBlock = (Type, marginBottom) => {
  const InnerBlock = ({ data, attribs }) => (
    <CopyLink attribs={attribs} marginBottom={marginBottom}>
      <Type {...attribs}>
        <Html data={data} />
      </Type>
    </CopyLink>
  );
  InnerBlock.propTypes = {
    data: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
    attribs: PropTypes.object,
  };
  return InnerBlock;
};

export default LinkableHtmlBlock;
