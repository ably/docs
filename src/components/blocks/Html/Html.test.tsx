import React from 'react';
import { render, screen } from '@testing-library/react';
import { DlWrapper } from 'src/components/blocks/list/Dl/DlWrapper';
import { PageLanguageContext } from 'src/contexts';

import Html from './';
import { dtFixture } from './fixtures';

describe('<Html />', () => {
  it('renders correct Dl based on PageLanguageContext value', () => {
    const { container } = render(
      <PageLanguageContext.Provider value="python">
        <Html data={dtFixture} BlockWrapper={DlWrapper} />
      </PageLanguageContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
