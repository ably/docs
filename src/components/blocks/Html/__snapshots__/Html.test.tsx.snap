// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Html /> renders correct Dl based on PageLanguageContext value 1`] = `
<div>
  
	
  <dt
    class="sc-hKMtZM hqbRVj"
  >
    name
  </dt>
  
	
  <dd
    class="sc-gsnTZi dJqSEs"
  >
    event name for the published message
    <span
      lang="python"
    >
      <em>
        Type: 
        <code
          class="font-mono font-semibold text-code p-4"
        >
          Unicode
        </code>
         for Python 2, 
        <code
          class="font-mono font-semibold text-code p-4"
        >
          String
        </code>
         for Python 3
      </em>
    </span>
  </dd>
  
	
  <div
    lang="python"
  >
    <dt
      class="sc-hKMtZM hqbRVj"
    >
      <div>
        data
      </div>
    </dt>
    <dd
      class="sc-gsnTZi dJqSEs"
    >
      data payload for the message. The supported payload types are unicode Strings, Dict, or List objects that can be serialized to 
      <span
        class="caps"
      >
        JSON
      </span>
       using 
      <code
        class="font-mono font-semibold text-code p-4"
      >
        json.dumps
      </code>
      , binary data as 
      <code
        class="font-mono font-semibold text-code p-4"
      >
        bytearray
      </code>
       (in Python 3, 
      <code
        class="font-mono font-semibold text-code p-4"
      >
        bytes
      </code>
       also works), and None.
      <em>
        Type: 
        <code
          class="font-mono font-semibold text-code p-4"
        >
          Object
        </code>
      </em>
    </dd>
  </div>
  
	
  
	
  
	
  
	
  
	
  

</div>
`;
